### ISSUE TITLE
FreeOTP predicting codes on gitlab ce servers

### ISSUE THREAD

**Issue by afwn90cj93201nixr2e1re** [freeotp/freeotp-android#237]

**Comment by afwn90cj93201nixr2e1re** [comment]
Seems like there's some time-based problems, due codes - real, but they are predicted, they are really from the future...
I didn't test it with something else, but just installed oauth and it's shown code that was active 40-80 seconds ago and it's worked for now.

**Comment by mohe2015** [comment]
It's totally normal that OTP codes from a minute ago or the next one are accepted as valid. This is so it works if the clocks aren't perfectly in sync of you are too slow to type the code. I hope I understood you correctly.

**Comment by afwn90cj93201nixr2e1re** [comment]
So, you wanna say that i should remember code, then wait 1 minute and then insert old code? That's not really nice, why then Oauth app doesn't have that issue? Maybe you should use some public mtp servers for time synchronization instead local phones time?

**Comment by afwn90cj93201nixr2e1re** [comment]
@dvanbalen

**Comment by mohe2015** [comment]
@afwn90cj93201nixr2e1re if gitlab CE doesn't accept the code it may be that their OTP implementation is not correct. So no you are not supposed to wait it is supposed to just work of the time difference isn't too large. You may check the time on the server and on your phone. It shouldn't be off by more than a minute or so. (At least as far as I know, I'm not using FreeOTP currently)

**Comment by afwn90cj93201nixr2e1re** [comment]
> To check if FreeOTP is at fault you could use Google Authenticator 

Bruh, are ....?
![изображение](https://user-images.githubusercontent.com/9306996/89122969-5350d280-d4bb-11ea-834d-c328f9dd0907.png)
Pls, start reading well.

FreeOTP can use MTP servers.

I don't have issues with that app, im just reporting, im not giving a damn about is that gonna be fixed or not, it's not rly necessary for me.

**Comment by mohe2015** [comment]
I read correctly but as you are on the bug reporter for freeotp I assumed you just miswrote the app name. 
Because the title is "FreeOTP predicting codes on gitlab ce servers" and then "I didn't test it with something else, but just installed oauth" I thought you meant FreeOTP by oauth. Also I don't know what oauth is for an app. I just know the authentication scheme called oauth which has nothing to do with OTP or FreeOTP.

Also here "That's not really nice, why then Oauth app doesn't have that issue? Maybe you should use some public mtp servers for time synchronization instead local phones time?" you are doing a comparison of something with the oauth app. What are you comparing it to?

Also why are you reporting it here if it doesn't have anything to do with FreeOTP?

People make strong assumptions and if those don't hold it gets messy (especially as I'm not a native speaker)

**Comment by afwn90cj93201nixr2e1re** [comment]
> Also why are you reporting it here if it doesn't have anything to do with FreeOTP?

????
> especially as I'm not a native speaker

im also
But u are right.
https://stackshare.io/stackups/authy-vs-oauth-io

Rn bug related ti freeotp exactly.

**Comment by mohe2015** [comment]
I'm really sorry but I don't think I can help you any more. It is really hard for me to follow this conversation. Maybe somebody else can take over. Also it may help in general to write step by step reproduction steps (very verbose, writing what exactly you are doing, at what time, in what order, step by step). Screenshots or videos usually also help. It's also totally fine to edit comments to improve their wording or so.

**Comment by bostrt** [comment]
Hi @afwn90cj93201nixr2e1re and @mohe2015, I've been watching this issue closely trying to identify a resolvable issue or bug. I tend to agree with you, @mohe2015, in that the report isn't coherent enough at this point to move forward with anything. 

@afwn90cj93201nixr2e1re, please take @mohe2015's advise and clearly re-state the issue along with steps on reproducing the problem.

**Comment by afwn90cj93201nixr2e1re** [comment]
Damn, just download app, try to auth with that in gitlab ce services, what else i should explain? I do nothing else.

**Comment by afwn90cj93201nixr2e1re** [comment]
Whatever, i just reported it, if you wanna - you can fix it, else that's not my problem, just add new checkbox to app and named it as Time sync Google mtp server for example, and if it's checked request current time from mtp server and use it for calculating instead local phone time.

**Comment by bostrt** [comment]
> Damn, just download app, try to auth with that in gitlab ce services, what else i should explain? I do nothing else.

I gave this a try earlier today following these instructions: https://about.gitlab.com/install/#centos-7. I used gitlab-ee-13.2.2-ee.0.el7.x86_64 and v1.5 FreeOTP on Android. I'm also using network provided time on my android (I assume you meant NTP and not MTP). 

This is all working fine for me so I recommend comparing the timestamps from your Android device and the server hosting your GitLab CE.
